pulpcore/plugin/repo_version_utils.py

```diff
@@ -31,21 +31,36 @@ def remove_duplicates(repository_version):
         repository_version: The :class:`~pulpcore.plugin.models.RepositoryVersion` to be checked
             and possibly modified.
     """
-    query_for_repo_duplicates_by_type = defaultdict(lambda: Q())
-    for item in repository_version.added():
-        detail_item = item.cast()
-        if detail_item.repo_key_fields == ():
+    added_content = repository_version.added().only('pk')
+    existing_content = repository_version.content.exclude(pk__in=added_content)
+    repository = repository_version.repository.cast()
+    content_types = {ctype.get_pulp_type(): ctype for ctype in repository.CONTENT_TYPES}
+
+    for pulp_type in content_types.keys():
```

Comment:
```suggestion
    for pulp_type, type_obj in content_types.items():
```

and remove the next line.